### ISSUE TITLE
project_git and project_mbox are alphabetically ordered and not by date in social_smell_showcase.Rmd

### ISSUE THREAD

**Issue by tuejari** [sailuh/kaiaulu#126]
`project_git` and `project_mbox` are alphabetically ordered and not by date.

Note that `project_git` and `project_mbox` are ordered almost directly after parsing as seen in the LOC below.
https://github.com/sailuh/kaiaulu/blob/65a99b40304af16d07470be8e9eac6b221251ceb/vignettes/social_smell_showcase.Rmd#L92-L96

https://github.com/sailuh/kaiaulu/blob/65a99b40304af16d07470be8e9eac6b221251ceb/vignettes/social_smell_showcase.Rmd#L104-L105

The datetime object is stored as "chr" and not formatted as YYYY-MM-DD HH:mm:ss Z. An example of the format that it uses is: 
```Fri Apr 1 09:58:42 2016 -0400```. This means that they are not ordered by date, but alphabetically. An example for `project_git` is shown in the picture below.

![image](https://user-images.githubusercontent.com/33750620/131363668-221d7491-c284-4ecd-8bd1-fdffc828df21.png)

First, the dates are formatted as yyyy-MM-DD hh:mm:ss which allows for correct ordering in the following lines:
https://github.com/sailuh/kaiaulu/blob/65a99b40304af16d07470be8e9eac6b221251ceb/vignettes/social_smell_showcase.Rmd#L124-L129

However, note that the identity matching sorts the table by `identity_id`. So `project_git` and `project_mbox` may need to be ordened after the following lines:
https://github.com/sailuh/kaiaulu/blob/65a99b40304af16d07470be8e9eac6b221251ceb/vignettes/social_smell_showcase.Rmd#L165-L174

I have done some testing and found that **_this bug affects the results_**.

I plan to make a pull request to fix this issue later.

**Comment by tuejari** [comment]
Actually, I am not 100% sure whether `project_git` and `project_mbox` need to be sorted by `identity_id` (primary) and the date (secondary) or only by date. However, I know for sure that it is currently reporting the wrong commit window in the resulting CSV since the first and last rows of `project_git` are taken:

https://github.com/sailuh/kaiaulu/blob/65a99b40304af16d07470be8e9eac6b221251ceb/vignettes/social_smell_showcase.Rmd#L260-L262

https://github.com/sailuh/kaiaulu/blob/65a99b40304af16d07470be8e9eac6b221251ceb/vignettes/social_smell_showcase.Rmd#L413-L416

https://github.com/sailuh/kaiaulu/blob/65a99b40304af16d07470be8e9eac6b221251ceb/vignettes/social_smell_showcase.Rmd#L458-L462

 @carlosparadis If you have time, could you please take a look at what the correct solution is here?

**Comment by carlosparadis** [comment]
Hi Jari,

Thank you for finding this. You are correct, this is being done before conversion to date. This issue was likely introduced when `parse_gitlog` stopped being in charge of also parsing the data dates and it was deferred to the notebook (as the quality framework metric num.tz would be impossible to calculate otherwise). At this point in the code where the sort happens, the timezones are also not "as-is" which could further mess any code reliant on the assumption the table is sorted.

With that being said:

> I have done some testing and found that this bug affects the results.

Could you elaborate if you believe anything besides 

> I know for sure that it is currently reporting the wrong commit window in the resulting CSV  

is currently being affected?

The code snipets you refer above:

```r
i_commit_hash <- data.table::first(project_git_slice)$commit_hash 
j_commit_hash <- data.table::last(project_git_slice)$commit_hash 
```

Are applied to project git slice, not project git (note this happens after the code snippet to date conversion). The process conceptually should go like this:

1) First the slices are identified in an array of slices on the `time_window` variable:

https://github.com/sailuh/kaiaulu/blob/65a99b40304af16d07470be8e9eac6b221251ceb/vignettes/social_smell_showcase.Rmd#L196-L208

`start_commit` and `end_commit` , which in turn are defined by the user on the project configuration file:

https://github.com/sailuh/kaiaulu/blob/65a99b40304af16d07470be8e9eac6b221251ceb/vignettes/social_smell_showcase.Rmd#L69-L70

So the range of time windows should be correct. 

2) The question then is whether the rows that go into each slice (which subsequently will be used to compute the metrics) is correct:

https://github.com/sailuh/kaiaulu/blob/65a99b40304af16d07470be8e9eac6b221251ceb/vignettes/social_smell_showcase.Rmd#L246-L262

Lines 246-256 are in charge of that, not 261 and 262. Lines 246-256 define the content of `project_git_slice` and `project_mbox_slice`. Since this is a boolean operator over the already converted timestamp, and not based on order, I would believe this to be correctly assigning the right rows to the right slices. However, the `first_commit` and the `last_commit` of said time slices are likely incorrect, since they are based on the position of the table. 

3) So if we agree the content that goes in each time slice is correct, but the first and last commit hashes can be wrong for depending on sorting, then the concern is when the slice is put together:

https://github.com/sailuh/kaiaulu/blob/65a99b40304af16d07470be8e9eac6b221251ceb/vignettes/social_smell_showcase.Rmd#L336-L352

The `commit_interval` is potentially incorrect due to the issue you noted, as it is a simple concatenation of `i_commit_hash` and `j_commit_hash`. However, `start_day` and `end_day` are obtained from the vector of slices I noted before, which should be correct:

https://github.com/sailuh/kaiaulu/blob/65a99b40304af16d07470be8e9eac6b221251ceb/vignettes/social_smell_showcase.Rmd#L241-L243

4) Finally, another point of concern that came to mind is whether these commits are used to align the data from various sources (e.g. Line Metrics with Social Smell Metrics):

https://github.com/sailuh/kaiaulu/blob/65a99b40304af16d07470be8e9eac6b221251ceb/vignettes/social_smell_showcase.Rmd#L498-L504

This is done by `commit_interval`, which is the variable of concern. This one is interesting since it is an inner join, so if there were inconsistencies with i_commit_hash and j_commit_hash for example, we would have caught much sooner entire time windows were just missing or the table would be empty. But I don't recollect any occurrence of that insofar. My guess is that since the subsetting code is, in essence, the same across Line Metrics section, Social Smell section etc (they could in fact be a single giant loop but readability suffers in the Notebook), I'm assuming the issue you raised affects all these loops systematically. 

In other words, the potentially incorrect i_commit_hash and j_commit_hash, if they are reported wrong for a time window in Social Smells, it will also consistently be incorrectly put for the Line Metrics. End result, the join will still be done "correctly" thanks to the consistency, even though the hash value itself **mismatches to the start_date and end_date**. 

5) TL;DR:

I'd expect the final result table to be correct, including the "start_date" and "end_date" reported in the table, but the `commit_interval` to be potentially incorrect (as it is subject to the order as you reported).

Assuming I am not overlooking anything and you agree with me on the above, I'd imagine two fixes needed to be done:

 a) 

https://github.com/sailuh/kaiaulu/blob/65a99b40304af16d07470be8e9eac6b221251ceb/vignettes/social_smell_showcase.Rmd#L96

https://github.com/sailuh/kaiaulu/blob/65a99b40304af16d07470be8e9eac6b221251ceb/vignettes/social_smell_showcase.Rmd#L105

These should be moved to the end of these code blocks respectively:

https://github.com/sailuh/kaiaulu/blob/65a99b40304af16d07470be8e9eac6b221251ceb/vignettes/social_smell_showcase.Rmd#L112-L120

https://github.com/sailuh/kaiaulu/blob/65a99b40304af16d07470be8e9eac6b221251ceb/vignettes/social_smell_showcase.Rmd#L123-L132

So it can properly sort by date and:

b) 

https://github.com/sailuh/kaiaulu/blob/65a99b40304af16d07470be8e9eac6b221251ceb/vignettes/social_smell_showcase.Rmd#L260-L262

These codes need to be replaced to obtain the actual first and last date. And in general, no code in this Notebook should be reliant on the row index (merges will destroy any sorting if not parameterized properly for optimization, so this is in general a terrible idea anyway).

So: 

```i_commit_hash <- data.table::first(project_git_slice)$commit_hash``` 

probably should be replaced with something like:

```    
i_commit_hash <- project_git[author_datetimetz == min(project_git$author_datetimetz,na.rm=TRUE)]$commit_hash
j_commit_hash <- project_git[author_datetimetz == max(project_git$author_datetimetz,na.rm=TRUE)]$commit_hash
```

Keep in mind the suggestion for a) is not to guarantee other code can use row-position operations, but rather in case someone wants to explore the project_git and project_mbox tables (e.g. run the code block up to the date conversion and sorting and call View()). 

Please let me know if you disagree, you are probably more familiar now with this code than I am. 😄 It would be great if you could do the PR, otherwise I will eventually get to it on a weekend.

@harismumtaz18 Assuming my understanding of the above is right, this should not affect your analysis since you have been relying on the date column rather than commit_hash. @valentina-lenarduzzi this issue may be relevant to you since you are working on this notebook.

**Comment by tuejari** [comment]
Hi Carlos,

Thank you for the elaborate response. 
I have done some more testing on a new branch with minimal differences to see if I had made a mistake somewhere else.

I have implemented 5a (from the TL;DR) as you stated. 

Regarding 5b:
You stated that  the assignments to `i_commit_hash` and `j_commit_hash` should be replaced with something like:
```r
i_commit_hash <- project_git[author_datetimetz == min(project_git$author_datetimetz,na.rm=TRUE)]$commit_hash
j_commit_hash <- project_git[author_datetimetz == max(project_git$author_datetimetz,na.rm=TRUE)]$commit_hash
```
I noticed that you used `project_git` here, not `project_git_slice`. Additionally, this function assigns an array of equal hashes to the variables (since commits occur multiple times in `project_git` and `project_git_slice`). I have replaced these lines with the following. I am not sure if this is the prettiest solution in R, but this is the best I could come up with for now.

```r
i_commit_hash <- data.table::first(project_git_slice[author_datetimetz == min(project_git_slice$author_datetimetz,na.rm=TRUE)])$commit_hash
j_commit_hash <- data.table::first(project_git_slice[author_datetimetz == max(project_git_slice$author_datetimetz,na.rm=TRUE)])$commit_hash
```

Below are my results (from before and after the fixes):

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/GEBRUI~1/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/GEBRUI~1/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
</head>

<body link="#0563C1" vlink="#954F72">



  | commit_interval | start_datetime | end_datetime | org_silo | missing_links | radio_silence | num_tz | code_only_devs | code_files | ml_only_devs | ml_threads | code_ml_both_devs | churn | n_commits | git_checkout | Lines | Code | Comments | Blanks | Complexity
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
Before fixes | be421a6b07c2ae9167150b77dc1185a94812b358-c90284d84e42993204d84cccaf5c03359ca0cdbd | 2021-04-22T14:28:03.532Z | 2021-07-20T22:54:49.532Z | 291 | 291 | 14 | 15 | 122 | 1127 | 76 | 147 | 14 | 0 | 405 | c90284d84e42993204d84cccaf5c03359ca0cdbd | 316858 | 240132 | 52643 | 24083 | 8248
After fixes | 4c8a32c8c58f165158d0fd36dcce55e05514d3d7-960da8a9074a4fb58881ea79f7dc9bc8fd58a5c4 | 2021-04-22T14:28:03.047Z | 2021-07-20T22:54:49.047Z | 321 | 321 | 14 | 15 | 122 | 1127 | 76 | 147 | 14 | 0 | 405 | 960da8a9074a4fb58881ea79f7dc9bc8fd58a5c4 | 323821 | 246695 | 52752 | 24374 | 8308



</body>

</html>

As you can see, I get different results for `org_silo`, `missing_links`, `Lines`, `Code`, `Comments`, `Blanks`, and `Complexity`.
I am not sure what is causing the difference in results, but I have not yet looked too deep into it (currently 2AM here). However, at least the commit interval is now correct.

**Comment by carlosparadis** [comment]
Hi,

Yes, thanks for replacing to slice and using first(), this should be ok since the return of multiple rows occurs when the commit has more than one file (on my quick test I guess I got a commit of a single file).

Concerning `Lines`, `Code`, `Comments`, `Blanks`, and `Complexity`, in retrospect, this makes sense to have discrepancies. I skipped to merge when checking, but I neglected to consider this:

https://github.com/sailuh/kaiaulu/blob/65a99b40304af16d07470be8e9eac6b221251ceb/vignettes/social_smell_showcase.Rmd#L464

For some tools we use, analysis has to be done based on "what is in the folder in a specified filepath". Kaiaulu works around that by performing `git_checkout()` over a commit hash. In which case, as shown above, `j_commit_hash` is used. Therefore...if there was a misalignment between the dates being used and commit hashes, which you verified in raising this issue, then the line metrics above are incorrectly associated to a different time window (that of the incorrect commit range). 

So yes, the analysis is therefore compromised for those metrics as the data you show reflects. This should also be the case on the social-smell branch, but not the case on any analysis done before the https://github.com/sailuh/kaiaulu/commit/4b844cc8c978a51f23dae540e3400847db998eb7.  

With that being said... I am not certain why `org_silo` and `missing_links` are being affected by it, as the git checkout workaround is not used on them (since Kaiaulu does not defer analysis to a 3rd party tool requiring the workaround). Neither uses the clustering algorithm either, and I believe I seeded it for reproducibility anyway. 

If we confirm `org_silo` and `missing_link` are being affected by the commit_hash, then @harismumtaz18 @valentina-lenarduzzi  and the analysis Selma is doing for @rnkazman will be affected by it.

**Comment by tuejari** [comment]
Hi,

Thank you for the explanation on why it makes sense that the other metrics are affected.

> With that being said... I am not certain why org_silo and missing_links are being affected by it, as the git checkout workaround is not used on them (since Kaiaulu does not defer analysis to a 3rd party tool requiring the workaround). Neither uses the clustering algorithm either, and I believe I seeded it for reproducibility anyway.

I have inspected the environment for both tests, but there are no obvious differences, except for the different values of the results.
To me it is also unclear why `org_silo` and `missing_links` are affected.

**Comment by carlosparadis** [comment]
@tuejari Did you modify by any chance the project configuration file after you executed the code? Just wondering if the fact you are looking at a single time window may trigger an edge case.

Alternatively, could you upload here your project configuration file and link me a drive link to the git log and mailing list you are using so I can try to reproduce it?

**Comment by tuejari** [comment]
> @tuejari Did you modify by any chance the project configuration file after you executed the code? Just wondering if the fact you are looking at a single time window may trigger an edge case.

I thought this might've been the case to, so for the second test (reported in this comment: (https://github.com/sailuh/kaiaulu/issues/126#issuecomment-908793138) I used a new branch with minimal differences to see if I had made a mistake somewhere else. I do not think this is the case.

I have send you a mail linking the drive with my test files so you can try to reproduce it.

**Comment by carlosparadis** [comment]
Thanks for sending your data. I can't reproduce the social smells mismatch (but the line metrics mismatch occurs as we discussed):

before fixing the commit_hash: 

commit_interval | start_datetime | end_datetime | org_silo | missing_links | radio_silence | num_tz | code_only_devs | code_files | ml_only_devs | ml_threads | code_ml_both_devs | churn | n_commits | git_checkout | Lines | Code | Comments | Blanks | Complexity
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
7f6ddda710dd701f112973e488dabfab4fc723de-6ae9aeec3f866ecfce83945ad1e831be9f8e5ebb | 2021-04-22 14:28:03 | 2021-07-20 22:54:49 | 288 | 288 | 15 | 15 | 130 | 1243 | 78 | 153 | 15 | 0 | 431 | 6ae9aeec3f866ecfce83945ad1e831be9f8e5ebb | 264058 | 183879 | 62419 | 17760 | 8800

after fixing the commit_hash:

commit_interval | start_datetime | end_datetime | org_silo | missing_links | radio_silence | num_tz | code_only_devs | code_files | ml_only_devs | ml_threads | code_ml_both_devs | churn | n_commits | git_checkout | Lines | Code | Comments | Blanks | Complexity
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
4c8a32c8c58f165158d0fd36dcce55e05514d3d7-3a7c7648c67e10bdaccb56e4f7714f500c37c8bc | 2021-04-22 14:28:03 | 2021-07-20 22:54:49 | 288 | 288 | 15 | 15 | 130 | 1243 | 78 | 153 | 15 | 0 | 431 | 3a7c7648c67e10bdaccb56e4f7714f500c37c8bc | 267866 | 187056 | 62795 | 18015 | 8872


@tuejari Your issue is on line https://github.com/sailuh/kaiaulu/blob/65a99b40304af16d07470be8e9eac6b221251ceb/vignettes/social_smell_showcase.Rmd#L486


However, I didn't try running your notebook but the one currently on master branch. Could you try to run that with your config, git log and mbox and see if you still have the same issue? Also for the git_checkout() I used `main` instead of `origin/main`.

**Comment by carlosparadis** [comment]
Edit: If you read my last message on e-mail, please ignore the "your issue is on line X". I originally thought you used trunk, but after seeing your code I realized you had that changed.

**Comment by tuejari** [comment]
Thank you for your attempt at reproducing. I notice that you specifically mention: 
> before fixing the commit_hash:
and
> after fixing the commit_hash:

In your test to reproduce the social smells mismatch, did you also fix where the `project_git` and `project_mbox` get ordered (see 5a from https://github.com/sailuh/kaiaulu/issues/126#issuecomment-908690585) and not just the commit_hash? I believe that the change in where the `project_git` and `project_mbox` get ordered is causing the difference in results. 

I have tried it again on the kaiaulu master branch using `main` for git_checkout() instead of `origin/main` and I do get the social smells mismatch:

before fixing the commit_hash and the ordering issue:


<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/GEBRUI~1/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/GEBRUI~1/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

</head>

<body link="#0563C1" vlink="#954F72">



commit_interval | start_datetime | end_datetime | org_silo | missing_links | radio_silence | num_tz | code_only_devs | code_files | ml_only_devs | ml_threads | code_ml_both_devs | churn | n_commits | git_checkout | Lines | Code | Comments | Blanks | Complexity
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
7f6ddda710dd701f112973e488dabfab4fc723de-6ae9aeec3f866ecfce83945ad1e831be9f8e5ebb | 22/04/2021 14:28 | 20/07/2021 22:54 | 288 | 288 | 15 | 15 | 130 | 1243 | 78 | 153 | 15 | 0 | 431 | 6ae9aeec3f866ecfce83945ad1e831be9f8e5ebb | 264058 | 183879 | 62419 | 17760 | 8800



</body>

</html>


after fixing the commit_hash and the ordering issue:


<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/GEBRUI~1/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/GEBRUI~1/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

</head>

<body link="#0563C1" vlink="#954F72">



commit_interval | start_datetime | end_datetime | org_silo | missing_links | radio_silence | num_tz | code_only_devs | code_files | ml_only_devs | ml_threads | code_ml_both_devs | churn | n_commits | git_checkout | Lines | Code | Comments | Blanks | Complexity
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
4c8a32c8c58f165158d0fd36dcce55e05514d3d7-3a7c7648c67e10bdaccb56e4f7714f500c37c8bc | 22/04/2021 14:28 | 20/07/2021 22:54 | 322 | 322 | 15 | 15 | 130 | 1243 | 78 | 153 | 15 | 0 | 431 | 3a7c7648c67e10bdaccb56e4f7714f500c37c8bc | 267866 | 187056 | 62795 | 18015 | 8872



</body>

</html>


For clarity I will send you an email with files showing the git diff before the fixes and after the fixes.

**Comment by carlosparadis** [comment]
Let me rectify what I said above: While the number of the rows was indeed the same, the content was not. Hence, I now believe the error lies within: 

https://github.com/sailuh/kaiaulu/blob/65a99b40304af16d07470be8e9eac6b221251ceb/vignettes/social_smell_showcase.Rmd#L286-L288

To be more precise, I re-run the code and compared the contents of the variables "network_mbox_slice" and "mbox_network_authors", returned from the function calls `transform_reply_to_bipartite_network()` and `bipartite_graph_projection()` respectively on:

https://github.com/sailuh/kaiaulu/blob/65a99b40304af16d07470be8e9eac6b221251ceb/vignettes/social_smell_showcase.Rmd#L283-L289

After sorting the table by rows, the content of `network_mbox_slice` appeared to be the same, but the `mbox_network_authors` was not when modifying the sort from alphabetical to temporal. 

Weirdly, the temporal sort seems to lead to incorrect results (i.e. I found one particular instance two developers shared the e-mail thread and the alphabetical sorted correctly link them, but the temporal sorted did not.

**Comment by tuejari** [comment]
Thank you for the updates and detailed explanation!

> Weirdly, the temporal sort seems to lead to incorrect results (i.e. I found one particular instance two developers shared the e-mail thread and the alphabetical sorted correctly link them, but the temporal sorted did not).

I want to make sure that I understand correctly. Please correct me if I'm wrong.
So currently we don't know why temporal sort leads to incorrect results, but we do have an example to show that it leads to incorrect results. However, that does not necessarily mean that the results from the alphabetical ordering are correct, right?

**Comment by carlosparadis** [comment]
Correct. I will send you the example by e-mail since it contains an actual person name and e-mail on the data.

**Comment by tuejari** [comment]
Thank you for the example, just wanted to let you know here that it makes sense to me now.

**Comment by carlosparadis** [comment]
I think the problem is now solved. As I said at the beginning, social smells shouldn't be affected by the sorting order. Here's what I get after the fix:


**Sorted Alphabetically:**

commit_interval | start_datetime | end_datetime | org_silo | missing_links | radio_silence | num_tz | code_only_devs | code_files | ml_only_devs | ml_threads | code_ml_both_devs | churn | n_commits | git_checkout | Lines | Code | Comments | Blanks | Complexity
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
4c8a32c8c58f165158d0fd36dcce55e05514d3d7-3a7c7648c67e10bdaccb56e4f7714f500c37c8bc | 2021-04-22 14:28:03 | 2021-07-20 22:54:49 | 680 | 700 | 15 | 15 | 130 | 1243 | 78 | 153 | 15 | 0 | 431 | 3a7c7648c67e10bdaccb56e4f7714f500c37c8bc | 267866 | 187056 | 62795 | 18015 | 8872

**Sorted Temporally:**

commit_interval | start_datetime | end_datetime | org_silo | missing_links | radio_silence | num_tz | code_only_devs | code_files | ml_only_devs | ml_threads | code_ml_both_devs | churn | n_commits | git_checkout | Lines | Code | Comments | Blanks | Complexity
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
4c8a32c8c58f165158d0fd36dcce55e05514d3d7-3a7c7648c67e10bdaccb56e4f7714f500c37c8bc | 2021-04-22 14:28:03 | 2021-07-20 22:54:49 | 680 | 700 | 15 | 15 | 130 | 1243 | 78 | 153 | 15 | 0 | 431 | 3a7c7648c67e10bdaccb56e4f7714f500c37c8bc | 267866 | 187056 | 62795 | 18015 | 8872

(Note in this before/after the fix for the commit hash is already applied, compared to the previous comparisons, so consequently, the Line Metrics also remain consistent).

The real issue was found in the social smells of org silo, missing link, and radio silence:

https://github.com/sailuh/kaiaulu/blob/65a99b40304af16d07470be8e9eac6b221251ceb/R/smells.R#L42-L43

The issue lines on line 43. 

Observe the textual names are mapped to a temporary id within function scope at this step. Now compare to this code portion of the same function:

https://github.com/sailuh/kaiaulu/blob/65a99b40304af16d07470be8e9eac6b221251ceb/R/smells.R#L53-L54

As you can see, only the `from` of `code.graph` was mapped to id, but not the `to`. Instead, mail.graph's got mapped instead. This means the function treated, through the analysis, half the edgelist as text ids and half as numeric ids. Considering even the same author could appear on either `from` or `to`, this would incorrectly assume the same person as different, and other operations such as `colab < vert` also become unclear since it is comparing numbers to text:

https://github.com/sailuh/kaiaulu/blob/65a99b40304af16d07470be8e9eac6b221251ceb/R/smells.R#L58

I have also noticed in missing link the mapping was missing, and the same in clusters, and have added code to address that. 


The good news is that this is a much more recent code change than the commit I mentioned earlier. So there is a chance your analysis may not be affected by this (e.g. I doubt @CorneJB was affected by it). 

More specifically, if you are using the `https://github.com/sailuh/kaiaulu/tree/88-add-social-smells` then you are using the old implementation of the function, which should be OK (at least in respect to this bug).

 --------------

With all that being said, the org silos and missing links are now more than double, so I want to do a bit more exploration on these fixes. I will push a fix this weekend if you want to give it a try since I'd imagine you are more familiar with this data at this point than I am.

**Comment by carlosparadis** [comment]
@tuejari could you test the changes pushed on https://github.com/sailuh/kaiaulu/commit/045355d3605a3fd778e6e7f630fd5798da638b8c and tell me if it resolves the issues raised here? Thank you a lot for finding these!

**Comment by tuejari** [comment]
> I think the problem is now solved. 

That is good news! Thank you so much for uncovering and solving the real issue causing the different results!

> Here's what I get after the fix

Interesting that org silos and missing links are now more than double. It seems a bit odd, as if they are counted double somehow.

> The real issue was found in the social smells of org silo, missing link, and radio silence:

Thank you for the detailed explanation on what was causing the issue. I thought that my initial idea that the ordering was affecting the results did not make sense either, I simply jumped to conclusions too quickly. It makes sense that the results are affected when it is comparing numbers to text.

> The good news is that this is a much more recent code change than the commit I mentioned earlier. So there is a chance your analysis may not be affected by this (e.g. I doubt @CorneJB was affected by it).

Since I had to postpone most of my analysis using Kaiaulu, I still have to do my analysis. I was lucky enough to notice the incorrect commit hashes in the results of the first community I applied Kaiaulu to for my analysis, and that my testing led me to identify the alphabetical ordering instead of only the use of the incorrect commit hash. 

> More specifically, if you are using the https://github.com/sailuh/kaiaulu/tree/88-add-social-smells then you are using the old implementation of the function, which should be OK (at least in respect to this bug).

I am not sure I understand this correctly, which function are you referring to here exactly? 
I was using the master branch for my analysis, but if I understand correctly, would we be able to compare the results of the old implementation to the results of the new implementation by switching between master and `https://github.com/sailuh/kaiaulu/tree/88-add-social-smells`? 

> @tuejari could you test the changes pushed on 045355d and tell me if it resolves the issues raised here? 

I have tried the changes pushed on 045355d once, but I did not have enough time yet to properly look at it in detail. I will comment on this later. Although I do get the same results as you.

> Thank you a lot for finding these!

No problem, I was lucky that I noticed the wrong commit hashes. Thank you very much for uncovering the exact problems and fixes!

**Comment by carlosparadis** [comment]
>> More specifically, if you are using the https://github.com/sailuh/kaiaulu/tree/88-add-social-smells then you are using the old implementation of the function, which should be OK (at least with respect to this bug).

> I am not sure I understand this correctly, which function are you referring to here exactly?
I was using the master branch for my analysis, but if I understand correctly, would we be able to compare the results of the old implementation to the results of the new implementation by switching between master and https://github.com/sailuh/kaiaulu/tree/88-add-social-smells?

This was more a remark to reassure whoever has been working on the analysis a while back, like @CorneJB, that their analysis didn't get affected by this issue on the smells (since they were using a different implementation). The functions on that branch rely on the igraph library and community detection algorithms too.

You wouldn't be able to compare to the old analysis, because the community detection algorithm used is different in radio silence (arguably better). Org silo and missing could be potentially comparable, but I no longer remember enough of the code to be certain. So I recommend sticking with the master branch moving forward. 


> Interesting that org silos and missing links are now more than double. It seems a bit odd, as if they are counted double somehow.

Yes, I think a good sanity check may be to consider the number of "potential organizational silos". For example, I saw that there are 115 developers on git log that never sent a reply on the mailing list in the project you are looking at, out of the total 190 developers in the system (numbers are off memory so maybe a few digits off). 

The maximum number of edges in such graph would be n(n-1)/2 => 6555. For each of these edges, they count as a org.silo smell if the pair of developers in it did not post any e-mail reply in the mailing list at all (whether with that person or someone else) in the time window. 

So we have 680/6555 (~10%) of every time two developers touched the same file in that particular time window, they never sent "any e-mail reply" in any e-mail thread. Which is a relatively low number still. 

If you have a chance, try looking at the org silo output (before the length() function is applied) to see if the output seems consistent with what you understand of the community too. I'm glad to hear we are at least getting the same numbers now.

@harismumtaz18 @valentina-lenarduzzi you will likely want to use this recent commit version, pending any further findings, instead of the older ones if you are using the `master` branch.

**Comment by carlosparadis** [comment]
I'm assuming this issue has been resolved with 045355d3605a3fd778e6e7f630fd5798da638b8c. Please feel free to re-open otherwise.
